File: 0155-min-stack-最小栈/0155.cpp
class MinStack {
    int stack[8192];
    int *p = stack;
    int minStack[8192] = {INT_MAX};
    int *minP = minStack;
public:
    /** initialize your data structure here. */
    MinStack() {
        
    }
    
    void push(int x) {
        *p = x;
        p++;
        *(minP+1) = *minP < x ? *minP : x;
        minP++; 
    }
    
    void pop() {
        p--;
        minP--;
    }
    
    int top() {
        return *(p-1);
    }
    
    int getMin() {
        return *minP;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack-最小栈/0155.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minStack = [float('inf')]
        

    def push(self, x: int) -> None:
        self.stack.append(x)
        if x < self.minStack[-1]:
            self.minStack.append(x)
        else:
            self.minStack.append(self.minStack[-1])
        
    def pop(self) -> None:
        self.stack.pop()
        self.minStack.pop()
        

    def top(self) -> int:
        return self.stack[-1]
        

    def getMin(self) -> int:
        return self.minStack[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()